### PR TITLE
Adding database functions + tests

### PR DESCRIPTION
**Pull request by AymericNoel**
Implementation of BalanceUpdater

### CODE REVIEW

File: src/Trakx.IndiceManager.Server.Tests/Unit/Data/BalanceUpdaterTest.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        public void TryUpdateUserBalance_should_validateMapping_when_sendedAmount_is_verificationAmount()
+        {
+            var retrievedUser = _daoCreator.GetRandomUserAddressDao(5);
+            retrievedUser.IsVerified.Should().BeFalse();
+            var transaction = new Transaction{Amount = 5};
+            _userAddressProvider.ValidateMappingAddress(default).ReturnsForAnyArgs(true);
+
+            _balanceUpdater.TryUpdateUserBalance(retrievedUser, transaction);
+            _userAddressProvider.ReceivedWithAnyArgs(1).ValidateMappingAddress(default);
+            _userAddressProvider.DidNotReceiveWithAnyArgs().UpdateUserBalance(default);
```
Comment (monsieurleberre) [inline src/Trakx.IndiceManager.Server.Tests/Unit/Data/BalanceUpdaterTest.cs:50]: but it should also update the balance to be the verified balance no? 🤔

File: src/Trakx.IndiceManager.Server/Data/BalanceUpdater.cs [hunk 2]
```diff
@@ -9,25 +12,43 @@ public interface IUserBalanceUpdater : IObserver<Transaction>
...
 
+
+        #region Implementation of IUserBalanceUpdater
         /// <inheritdoc />
-        public bool TryUpdateUserBalance(Transaction transaction)
+        public bool TryUpdateUserBalance(IUserAddress retrievedUser, Transaction transaction)
         {
-            //database update ... 
-            return false;
+            var retrievedUserDao = (UserAddressDao)retrievedUser;
```
Comment (monsieurleberre) [inline src/Trakx.IndiceManager.Server/Data/BalanceUpdater.cs:35]: No no, please don't do that, this is illegal.

File: src/Trakx.Persistence/DAO/UserAddressDao.cs [hunk 3]
```diff
@@ -5,19 +5,22 @@ namespace Trakx.Persistence.DAO
 {
     public class UserAddressDao : IUserAddress
     {
```
Comment (monsieurleberre) [inline src/Trakx.Persistence/DAO/UserAddressDao.cs:7]: Maybe let's add a creation date too please

File: src/Trakx.Tests/Unit/Persistence/UserAddressProviderTest.cs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            userAddressToSave.Balance += 20;
+            await _userAddressProvider.UpdateUserBalance(userAddressToSave);
+
+            var retrievedUserAddress =
+                await _context.UserAddresses.FirstOrDefaultAsync(u => u.Address == userAddressToSave.Address);
+            retrievedUserAddress.Balance.Should().Be(20);
+        }
+
+        [Fact]
+        public async Task AddNewMapping_should_if_mapping_not_in_database_yet()
```
Comment (monsieurleberre) [inline src/Trakx.Tests/Unit/Persistence/UserAddressProviderTest.cs:77]: should_(create_new_mapping)_If....

File: src/Trakx.Persistence/UserAddressProvider.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            if (retrievedUser != null)
+            {
+                if (retrievedUser.IsVerified)
+                    return false;
+
+                _dbContext.Entry(retrievedUser).Entity.VerificationAmount=userAddressToSave.VerificationAmount;
+                _dbContext.Entry(retrievedUser).Entity.UserId = userAddressToSave.UserId;
+                return await _dbContext.SaveChangesAsync() > 0;
+            }
+            await _dbContext.UserAddresses.AddAsync((UserAddressDao)userAddressToSave);
```
Comment (monsieurleberre) [inline src/Trakx.Persistence/UserAddressProvider.cs:49]: this too :)

File: src/Trakx.IndiceManager.Server.Tests/Unit/Data/BalanceUpdaterTest.cs (hunk 1 above)
Comment (AymericNoel) [reply src/Trakx.IndiceManager.Server.Tests/Unit/Data/BalanceUpdaterTest.cs:50]: In my code, when you verified a User, you change the IsVerified status but you don't change the balance directly.
I do it in a database function.

Comment (monsieurleberre) [reply src/Trakx.IndiceManager.Server.Tests/Unit/Data/BalanceUpdaterTest.cs:48]: So we don't want to give them back what they gave us? I think we do, maybe there is something I don't understand...

File: src/Trakx.Common/Core/UserAddress.cs [hunk 6]
```diff
@@ -16,6 +16,7 @@ public UserAddress(string chainId, string userId, string address, decimal verifi
             IsVerified = isVerified;
             Balance = balance;
             Id = this.GetId();
+            CreationDate = creationDate;
```
Comment (monsieurleberre) [inline src/Trakx.Common/Core/UserAddress.cs:19]: thanks!

File: src/Trakx.IndiceManager.Server/Data/BalanceUpdater.cs [hunk 7]
```diff
@@ -32,7 +32,7 @@ public UserBalanceUpdater(IUserAddressProvider userAddressProvider)
         /// <inheritdoc />
         public bool TryUpdateUserBalance(IUserAddress retrievedUser, Transaction transaction)
         {
-            var retrievedUserDao = (UserAddressDao)retrievedUser;
+            var retrievedUserDao = new UserAddressDao(retrievedUser);
```
Comment (monsieurleberre) [inline src/Trakx.IndiceManager.Server/Data/BalanceUpdater.cs:35]: yes!

File: src/Trakx.IndiceManager.Server.Tests/Unit/Data/BalanceUpdaterTest.cs (hunk 1 above)
Comment (AymericNoel) [reply src/Trakx.IndiceManager.Server.Tests/Unit/Data/BalanceUpdaterTest.cs:48]: I do that here : ``` public async Task<bool> ValidateMappingAddress(IUserAddress userAddressToValidate)
        {
            var retrievedUser =
                await _dbContext.UserAddresses.FirstOrDefaultAsync(i => i.Address == userAddressToValidate.Address);

            _dbContext.Entry(retrievedUser).Entity.IsVerified = true;
            _dbContext.Entry(retrievedUser).Entity.Balance += retrievedUser.VerificationAmount;
            _dbContext.Entry(retrievedUser).Entity.VerificationAmount = 0;
            return await _dbContext.SaveChangesAsync() > 0;
        }```

File: src/Trakx.Persistence/UserAddressProvider.cs [hunk 8]
```diff
@@ -25,6 +25,8 @@ public async Task<bool> ValidateMappingAddress(IUserAddress userAddressToValidat
                 await _dbContext.UserAddresses.FirstOrDefaultAsync(i => i.Address == userAddressToValidate.Address);
 
             _dbContext.Entry(retrievedUser).Entity.IsVerified = true;
+            _dbContext.Entry(retrievedUser).Entity.Balance += retrievedUser.VerificationAmount;
+            _dbContext.Entry(retrievedUser).Entity.VerificationAmount = 0;
```
Comment (monsieurleberre) [inline src/Trakx.Persistence/UserAddressProvider.cs:29]: OK then I am not sure we want to reset this to 0. Is it going to be used for anything after that? If not I would rather keep it like that.

Comment (AymericNoel) [reply src/Trakx.Persistence/UserAddressProvider.cs:29]: hum, I will change that !

**Pull request merged**